### PR TITLE
Create search history feature

### PR DESCRIPTION
**Pull request by sashapff**
You don't need t review it now, just to see what I've done.
I will ask you a review when finish correct saving. Now it saves each request with duplicates and returns whole list of songs.
Go to `/history` to see this list.

### CODE REVIEW

**Comment by sashapff** [comment]
I fixed issues and now you can see the result on `/history` page

File: src/main/java/com/google/alpollo/model/SearchHistory.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+import java.util.ArrayList;
+import java.util.Comparator;
+import com.googlecode.objectify.annotation.Index;
+import com.googlecode.objectify.annotation.Id;
+import com.googlecode.objectify.annotation.Entity;
+import com.googlecode.objectify.annotation.Stringify;
+import com.google.alpollo.database.SongStringifier;
+
+/** 
+ * Represents the structure of user's searc history.
```
Comment (RazvanBudacaTechquarter) [inline src/main/java/com/google/alpollo/model/SearchHistory.java]: Typo here.

File: src/main/java/com/google/alpollo/model/SearchHistory.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  private static final int TOP_SIZE = 10;
+
+  /** Objectify requires no argument constructor. Do not use it. */
+  private SearchHistory() { }
+
+  public SearchHistory(String userId) {
+    id = userId;
+  }
+
+  /** Saves new  search request from autorized user. */
```
Comment (RazvanBudacaTechquarter) [inline src/main/java/com/google/alpollo/model/SearchHistory.java]: Extra space before `search` and authorize*

File: src/main/java/com/google/alpollo/model/SearchHistory.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  public void addSearchRequest(Song song) {
+    if (history.containsKey(song)) {
+      history.put(song, history.get(song) + 1);
+    } else {
+      history.put(song, 1);
+    }
+  }
+
+  /** Returns list with top 10 searched songs from autorized user. */
```
Comment (RazvanBudacaTechquarter) [inline src/main/java/com/google/alpollo/model/SearchHistory.java]: Authorized*

File: src/main/java/com/google/alpollo/model/SearchHistory.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+  public List<Song> getHistory() {
+    // sort by values and then add only top 10 songs to result list
```
Comment (RazvanBudacaTechquarter) [inline src/main/java/com/google/alpollo/model/SearchHistory.java]: Let's start with a capital S here.

File: src/main/java/com/google/alpollo/model/SearchHistory.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+      history.put(song, 1);
+    }
+  }
+
+  /** Returns list with top 10 searched songs from autorized user. */
+  public List<Song> getHistory() {
+    // sort by values and then add only top 10 songs to result list 
+    List<Song> sortedSongs = new ArrayList<>();
+   history.entrySet()
+       .stream()
```
Comment (RazvanBudacaTechquarter) [inline src/main/java/com/google/alpollo/model/SearchHistory.java]: Spacing seems a bit funky here.

File: src/main/java/com/google/alpollo/model/Song.java [hunk 6]
```diff
@@ -2,14 +2,16 @@
...
-/** 
- * Represents the structure of the song. 
+/**
+ * Represents the structure of the song.
  */
 public class Song {
-  @Index private String artist;
-  @Index private String name;
-  /** For counting hashcode. **/
+  @Index private String artist = "";
```
Comment (jeinygroove) [inline src/main/java/com/google/alpollo/model/Song.java]: why do you need default values here?

File: src/main/java/com/google/alpollo/model/Song.java [hunk 7]
```diff
@@ -19,6 +21,15 @@ public Song(String artist, String name) {
     this.name = name;
   }
 
+  /** For objectify and saving maps. */
```
Comment (jeinygroove) [inline src/main/java/com/google/alpollo/model/Song.java]: can you provide more info, please? why do we need this method and how objectify use it

File: src/main/java/com/google/alpollo/model/Song.java [hunk 8]
```diff
@@ -33,4 +44,9 @@ public String getName() {
   public Long id() {
     return BIG_PRIME_NUMBER * artist.hashCode() + name.hashCode();
   }
+
+  /** For objectify and saving maps. */
```
Comment (jeinygroove) [inline src/main/java/com/google/alpollo/model/Song.java]: can you provide more info, please? why do we need this method and how objectify use it

**Review by jeinygroove — COMMENTED** [review]
You accidentally deleted AnalysisTest.java :(

**Review by jeinygroove — APPROVED** [review]

File: src/main/java/com/google/alpollo/model/Song.java [hunk 9]
```diff
@@ -19,6 +21,18 @@ public Song(String artist, String name) {
     this.name = name;
   }
 
+  /** 
+   * Objectify requires to write this constructor to save maps to database.
+   * More detail we need to convert class Song to String and convert back because we use Song class as a keys.
```
Comment (jgaillard85) [inline src/main/java/com/google/alpollo/model/Song.java]: nit: key, not keys

File: src/main/java/com/google/alpollo/model/SongCounter.java [hunk 10]
```diff
@@ -32,4 +32,8 @@ public void incrementSearchCounter() {
   public int getSearchCounter() {
     return searchCounter;
   }
+
+  public void setSearchCounter(int searchCounter) {
```
Comment (jgaillard85) [inline src/main/java/com/google/alpollo/model/SongCounter.java]: I don't see where you are using this method

File: src/main/java/com/google/alpollo/model/SearchHistory.java [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+
+/** 
+ * Represents the structure of user's search history.
+ * Contains the list with all songs.
+ */
+@Entity
+public class SearchHistory {
+  @Id private String id;
+  @Stringify(SongStringifier.class)
+  private HashMap<Song, Integer> history = new HashMap<>();
```
Comment (jgaillard85) [inline src/main/java/com/google/alpollo/model/SearchHistory.java:23]: Add javadoc to explain what this map is storing

File: src/main/java/com/google/alpollo/database/SongDataBase.java [hunk 12]
```diff
@@ -49,4 +53,28 @@ public static void saveAnalysisInfo(AnalysisInfo info) {
...
+      searchHistory.addSearchRequest(song);
+      OfyService.ofy().save().entity(searchHistory).now();
+    } 
+  }
+
+  /** Returns the list of the most requested songs from the user. */
+  public static List<Song> searchHistory(String userId) {
+    SearchHistory searchHistory = OfyService.ofy().load().type(SearchHistory.class).id(userId).now();
+    if (searchHistory == null) {
+      return new ArrayList<>();
```
Comment (jgaillard85) [inline src/main/java/com/google/alpollo/database/SongDataBase.java]: Use Collections.emptyList() here, as that saves on creating a new object

**Review by jgaillard85 — APPROVED** [review]

**Pull request merged**